paint.js

```diff
@@ -10,68 +6,154 @@ var enableForce = false;
+        });
+        ["touchstart", "touchmove", "touchend"].forEach(function(e) {
+            elem.addEventListener(e, TouchHandler);
+        });
     }
-  }
 
-});
+    window.addEventListener("resize", function(e) {
+        InitializeCanvas();
+    });
+
+    document.addEventListener("keyup", function(e) {
+        switch(e.which) {
+        // ESC
+        case 27:
+            var canvas = document.getElementById("canvas");
+            canvas.getContext("2d").clearRect(0, 0, canvas.width, canvas.height);
+            break;
+
+        // p
+        case 80:
+            pointMode = !pointMode;
+            window.location.hash = pointMode ? "#points" : "";
+            break;
+
+        // f
+        case 70:
+            enableForce = !enableForce;
+            break;
+
+        // a
+        case 65:
+            drawTouchMajor = !drawTouchMajor;
+            break;
+
+        // enter
+        case 13:
+            if (document.documentElement.webkitRequestFullscreen) {
+                if (document.webkitFullscreenElement)
+                    document.webkitCancelFullScreen();
+                else
+                    document.documentElement.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
+            }
+        }
+    });
+}
 
-window.addEventListener("resize", function(e) {
-    InitializeCanvas();
-});
+function InitializeCanvas() {
+    var elem = document.getElementById("canvas");
+
+    var newscale = window.devicePixelRatio ? window.devicePixelRatio : 1;
+    var newwidth = window.screen.width * newscale;
+    var newheight = window.screen.height * newscale;
+
+    if (elem.width != newwidth || elem.height != newheight || scale != newscale) {
+        // resizing a canvas clears it, so do it only when it's dimensions have changed.
+        scale = newscale;
+        elem.width = newwidth;
+        elem.height = newheight;
+        elem.style.width = window.screen.width + "px";
+        elem.style.height = window.screen.height + "px";
+    }
+}
+
+function PointerHandler(event) {
+    event.preventDefault();
+
+    if (event.type == "pointerdown" && event.button === 0)
+        mousePressed = true;
+
+    if (mousePressed && (event.type == "pointerup" || (event.buttons & 1))) {
+        var fakeTouch = {
+            identifier: event.pointerId,
+            pageX : event.pageX,
+            pageY : event.pageY,
+            radiusX : event.width,
+            radiusY : event.height,
+            force : event.pressure,
+        };
+        var eventType = event.type == "pointerdown" ? "touchstart" :
+            event.type == "pointerup" ? "touchend" : "touchmove";
+        drawTouch(fakeTouch, eventType);
+        event.preventDefault();
+    }
+
+    if (event.type == "pointerup")
+        mousePressed = false;
+}
+
+function MouseHandler(event) {
+    if (event.type == "mousedown" && event.button === 0)
+        mousePressed = true;
+
+    if (mousePressed && event.button === 0) {
+        var fakeTouch = {
```

Comment:
Ugh, I hate that I wrote this this way rather than having all the events share common event-agnostic code ;-).